examples/main/main.cpp

```diff
@@ -331,6 +334,17 @@ int main(int argc, char ** argv) {
     while ((n_remain != 0 && !is_antiprompt) || params.interactive) {
         // predict
         if (embd.size() > 0) {
+            auto max_embd_size = n_ctx - 4;
```

Comment:
It comes from here: https://github.com/ggerganov/llama.cpp/blob/f5a790f761a43a0a7cbb7b55b76d4d7196ff2ea8/examples/main/main.cpp#L188

I can't tell you why it's 4 though. I started off using - 2 which seemed to work, but then I figured it was probably a bad idea using a different value than the logic that already existed. (- 1 makes sense to me because of the start of document token that's always added but that actually didn't work.)

Reply:
It looks like this was introduced when infinite generation was added in this commit, :
https://github.com/ggerganov/llama.cpp/commit/e2d490dafd860eaaaf9aa8008ab790527d556daf#diff-2d3599a9fad195f2c3c60bd06691bc1815325b3560b5feda41a91fa71194e805
So I guess it is related to that. At some point we should improve the documentation of the examples, but it is out of scope of this PR.

Reply:
I definitely wouldn't have any problem adding some comments, the issue here I don't really know the reasoning behind those decisions so I can't document them. I pretty much just assumed whoever wrote the stuff initially had a reason and tried to do basically the same thing.

I assume it's the case, but just double checking: Do you think this is good to merge without further changes?

Reply:
I think a comment explaining that the value was taken from the other line would be good, but the git blame will lead future reviewers to this PR, so I don't consider it critical.